ufl/algorithms/formtransformations.py

```diff
@@ -362,7 +359,18 @@ def compute_form_lhs(form):
         a = u*v*dx + f*v*dx
         a = lhs(a) -> u*v*dx
     """
-    return compute_form_with_arity(form, 2)
+    parts = tuple(sorted(set(part for a in form.arguments() if (part := a.part()) is not None)))
+    # If Arguments are stemming from a MixedFunctionSpace, we have to compute this per block
+    if parts == ():
+        return compute_form_with_arity(form, 2)
+
+    form_blocks = extract_blocks(form, arity=2)
+    lhs = 0
+    for bi in form_blocks:
+        for bj in bi:
+            if bj is not None:
+                lhs += compute_form_with_arity(bj, 2)
+    return lhs
```

Comment:
In the test above, we want to extract the blocks of the left and right hand sides of our form. What it does : 
(1) `lhs(F)` extracts the blocks and sum the corresponding subforms : `lhs += compute_form_with_arity(bj, 2)` 
(2) we do `a_blocked = extract_blocks(a)` where `a=lhs(F)`, meaning re-doing what the block extraction already implemented in `lhs()`
Just an idea: Would it be worth having functions `lhs_blocks()` and `rhs_block()`, or maybe some option to `lhs()` and `rhs()` to return the list of blocks and not the sum ?

Reply:
I don't think so, as on blocked structures (lists or nested lists), UFL cant perform any symbolical magic. Therefore, we want the final step of the algorithm before calling anything in DOLFINx to be `ufl.extract_blocks`.